test/hotspot/jtreg/runtime/Monitor/libMonitorWithDeadObjectTest.c

```diff
@@ -0,0 +1,172 @@
+  jclass clazz = (*env)->FindClass(env, "java/lang/System");
+  check(env, clazz, "No class");
+
+  jmethodID method = (*env)->GetStaticMethodID(env, clazz, "gc", "()V");
+  check(env, method, "No method");
+
+  (*env)->CallStaticVoidMethod(env, clazz, method);
+  check_exception(env, "Calling System.gc()");
+}
+
+static void thread_dump_with_locked_monitors(JNIEnv* env) {
+  jclass ManagementFactoryClass = (*env)->FindClass(env, "java/lang/management/ManagementFactory");
+  check(env, ManagementFactoryClass, "No ManagementFactory class");
+
+  jmethodID getThreadMXBeanMethod = (*env)->GetStaticMethodID(env, ManagementFactoryClass, "getThreadMXBean", "()Ljava/lang/management/ThreadMXBean;");
+  check(env, getThreadMXBeanMethod, "No getThreadMXBean method");
+
+  jobject threadBean = (*env)->CallStaticObjectMethod(env, ManagementFactoryClass, getThreadMXBeanMethod);
+  check(env, threadBean, "Calling getThreadMXBean()");
+
+  jclass ThreadMXBeanClass = (*env)->FindClass(env, "java/lang/management/ThreadMXBean");
+  check(env, ThreadMXBeanClass, "No ThreadMXBean class");
+
+  jmethodID dumpAllThreadsMethod = (*env)->GetMethodID(env, ThreadMXBeanClass, "dumpAllThreads", "(ZZ)[Ljava/lang/management/ThreadInfo;");
+  check(env, dumpAllThreadsMethod, "No dumpAllThreads method");
+
+  // The 'lockedMonitors == true' is what causes the monitor with a dead object to be examined.
+  jobject array = (*env)->CallObjectMethod(env, threadBean, dumpAllThreadsMethod, JNI_TRUE /* lockedMonitors */, JNI_FALSE /* lockedSynchronizers*/);
+  check(env, array, "Calling dumpAllThreads(true, false)");
+}
+
+static void create_monitor_with_dead_object(JNIEnv* env) {
+  jobject obj = create_object(env);
+
+  if ((*env)->MonitorEnter(env, obj) != 0) die("MonitorEnter");
+
+  // Drop the last strong reference to the object associated with the monitor.
+  // The monitor only keeps a weak reference to the object.
+  (*env)->DeleteLocalRef(env, obj);
+
+  // Let the GC clear the weak reference to the object.
+  system_gc(env);
+}
+
+static void* create_monitor_with_dead_object_in_thread() {
+  JNIEnv* env;
+  int res = (*jvm)->AttachCurrentThread(jvm, (void**)&env, NULL);
+  if (res != JNI_OK) die("AttachCurrentThread");
+
+  // Make the correct incantation to create a monitor with a dead object.
+  create_monitor_with_dead_object(env);
+
+  // DetachCurrenThread will try to unlock held monitors. This has been a
+  // source of at least two bugs:
+  // - When the object reference in the monitor was cleared, the monitor
+  //   iterator code would skip it, preventing it from being unlocked when
+  //   the owner thread detached, leaving it lingering in the system.
+  // - When the monitor iterator API was rewritten the code was changed to
+  //   assert that we didn't have "owned" monitors with dead objects. This
+  //   test provokes that situation and that asserts.
+  if ((*jvm)->DetachCurrentThread(jvm) != JNI_OK) die("DetachCurrentThread");
+
+  return NULL;
```

Comment:
Why is this function return type "void*" when it only returns NULL?

Reply:
@dcubed-ojdk These are the thread routines passed to `pthread_create` and must be typed as `void *(*start_routine)(void*)`.

Reply:
Though I just noticed the parameter is missing. As @stefank has pointed out this was copied from another test so all of that other test's issues are/were also present here unfortunately.

Reply:
I've updated the functions to use the correct signature.